### ISSUE TITLE
Error with minimal forward_step function

### ISSUE THREAD

**Issue by flixxox** [rwth-i6/returnn#1385]
Hi all,

I am using the torch backend and the `forward_step` function

```
def forward_step(*, model: Model, extern_data: TensorDict, **_kwargs):
    """
    Function used in inference.
    """
    init_forward_step_run_ctx(step=1)
    src = extern_data["data"].raw_tensor
    rf.get_run_ctx().mark_as_output(src, 'output')
```

gives me this error

```
File "/home/fschmidt/code/returnn/returnn/__main__.py", line 476, in execute_main_task
    line: engine.forward_with_callback(dataset=data, callback=forward_callback)
    locals:
      engine = <global> <returnn.torch.engine.Engine object at 0x7f3235717dd0>
      engine.forward_with_callback = <global> <bound method Engine.forward_with_callback of <returnn.torch.engine.Engine object at 0x7f3235717dd0>>
      dataset = <not found>
      data = <local> <TranslationDataset 'dataset_id139853662263952' epoch=1>
      callback = <not found>
      forward_callback = <local> <__returnn_config__.MyForwardCallback object at 0x7f3237902590>
  File "/home/fschmidt/code/returnn/returnn/torch/engine.py", line 734, in Engine.forward_with_callback
    line: model_outputs_per_batch.data[k] = _get_tensor_wo_batch_numpy(v)
    locals:
      model_outputs_per_batch = <local> TensorDict({})
      model_outputs_per_batch.data = <local> {}
      k = <local> 'output', len = 6
      _get_tensor_wo_batch_numpy = <local> <function Engine.forward_with_callback.<locals>._get_tensor_wo_batch_numpy at 0x7f3230121440>
      v = <local> Tensor{'output', ['const_dim0'(66),'const_dim1'(245)], dtype='int32'}
  File "/home/fschmidt/code/returnn/returnn/torch/engine.py", line 677, in Engine.forward_with_callback.<locals>._get_tensor_wo_batch_numpy
    line: if x.dims.index(batch_dim) != 0:
    locals:
      x = <local> Tensor{'output', ['const_dim0'(66),'const_dim1'(245)], dtype='int32'}
      x.dims = <local> (Dim{'const_dim0'(66)}, Dim{'const_dim1'(245)})
      x.dims.index = <local> <built-in method index of tuple object at 0x7f323028ae80>
      batch_dim = <local> Dim{B}
ValueError: tuple.index(x): x not in tuple
```

Does somebody know why?

**Comment by albertz** [comment]
Why do you call `init_forward_step_run_ctx(step=1)`? This is wrong. This function should never be called from within `forward_step`. This will likely cause all kinds of weird/wrong behavior.

But this is not the reason for your exception here. The exception says that it does not find the batch dim in the output. Specifically, it does not find the predefined `batch_dim` dim tag.

There are multiple ways how `mark_as_output` can be called:

The easiest way here in this example, and in general when you use the RETURNN frontend (RF, `returnn.frontend`), is by just passing a `returnn.tensor.Tensor` to it, i.e.:
```python
def forward_step(*, model: Model, extern_data: TensorDict, **_kwargs):
    src = extern_data["data"]
    rf.get_run_ctx().mark_as_output(src, 'output')
```
`src` already is using the `batch_dim` dim tag.

When you only have a pure `torch.Tensor`, you can explicitly wrap it into a `returnn.tensor.Tensor`, in multiple ways:

- `rf.convert_to_tensor(...)`
- `Tensor(...)`

In all cases, you need to define the dim tags for the resulting tensor.

Another way for `mark_as_output` is to define `model_outputs` in the config, in a similar way as you define `extern_data`. This will implicitly also define the dim tags of the outputs, and specifically it will also use `batch_dim` (depending how you define it). Then you can do:
```python
def forward_step(*, model: Model, extern_data: TensorDict, **_kwargs):
    src = extern_data["data"].raw_tensor
    rf.get_run_ctx().mark_as_output(src, 'output')
```

**Comment by flixxox** [comment]
> Why do you call init_forward_step_run_ctx(step=1)? This is wrong. This function should never be called from within forward_step. This will likely cause all kinds of weird/wrong behavior.

The function description tells me that I should call it :sweat_smile: 

```
def init_forward_step_run_ctx(*, expected_outputs: Optional[TensorDict] = None, step: Union[int, Tensor]):
    """
    Call this at the beginning of a new forward step.
    """
    global _run_ctx
    _run_ctx = RunCtx(stage="forward_step", expected_outputs=expected_outputs, step=step)
```

Thanks, converting to a returnn Tensor solved the issue:

```
def forward_step(*, model: Model, extern_data: TensorDict, **_kwargs):
    src = extern_data["data"].raw_tensor
    src = rf.convert_to_tensor(src, 
        dims=(
            batch_dim,
            Dim(None)
        )
    )
    rf.get_run_ctx().mark_as_output(src, 'output')
```

**Comment by albertz** [comment]
> Call this at the beginning of a new forward step.

This is maybe unclear. It means this is supposed to be called *before* the forward step. This is when you want to write your own custom engine, your own loop. I changed the documentation to make this more clear.

I also now changed the logic in `mark_as_output` on raw tensors, that it would actually allow to do what you did. It automatically assumes that the first dim is the batch dim.

I also improved the error message a bit.
